### PR TITLE
Add alert to notify about duplicate sample/metric ingestion

### LINKED ISSUE #1687: Alert and runbook for duplicate samples
This issue is based on a discusion in [#promscale slack thread](https://timescaledb.slack.com/archives/C011FC0PPC5/p1665473308182439?thread_ts=1664949014.661779&cid=C011FC0PPC5).

Promscale logs duplicate samples for reasons like:
1. Prometheus retrying a batch
2. Wrongly configured HA cluster

In both the above cases, poor ingestion throughput is the impact. We should alert the user about this early to limit the effect on ingestion and recommend necessary mitigation steps in the runbooks.

### PR DESCRIPTION
**Pull request by arajkumar**
## Description

This PR does the following,
1. Merge duplicate reporter into throughput reporter
2. Add alert about duplicate sample/metric ingestion
3. Add an e2e test to verify metrics related to duplicates are populated
4. Rename gauge metrics which ends with `total` to make linter happy [1]

[1] https://github.com/grafana/dashboard-linter/blob/44d415fb6bdc4d8e6585e514c448174d4de1ff02/lint/rule_target_counter_agg.go#L30

Signed-off-by: Arunprasad Rajkumar <ar.arunprasad@gmail.com>

Fixes #1687

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [X] CHANGELOG entry for user-facing changes
- [ ] Updated the relevant documentation

### CODE REVIEW

File: docs/runbooks/PromscaleIngestHighDataDuplication.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+Client payload has either duplicates or retrying many a time for the data which has been already ingested.
+
+
+## Impact
+
+Ingestion performance will be poor
+
+
+## Diagnosis
+1. If Prometheus is running in HA mode, go to [Prometheus high availability](#prometheus-high-availability)
```
Comment (alejandrodnm) [inline docs/runbooks/PromscaleIngestHighDataDuplication.md:14]: WDYT about an specific entry in diagnosis for prometheus HA running without the labels:

When running a Prometheus HA deployment there's an increase in ingest due to duplicate  data being sent to Promscale. We can view the ingest of duplicates with:

```
sum by(job, instance, type) (
          rate(promscale_ingest_duplicates_total{kind="sample"}[5m])
      )
```. 

This could happen if the Prometheus HA deployment is not configured to decorate the samples with the metadata from the replica that's pushing the data. In this scenario, two or more Prometheus replicas from the same cluster will be sending the exact same datapoints, and since there's no cluster/replica metadata, Promscale doesn't have the information needed to just accept the data from one of them and will try to persist them all.

File: pkg/pgmodel/ingestor/duplicates.go [hunk 2]
```diff
@@ -1,48 +0,0 @@
...
-	duplicateMetricsTotal uint64
-)
-
-func init() {
-	atomic.StoreUint64(&duplicateMetricsTotal, 0)
-}
-
-func registerDuplicates(duplicateSamples int64) {
-	metrics.IngestorDuplicates.With(prometheus.Labels{"type": "metric", "kind": "sample"}).Add(float64(duplicateSamples))
-	metrics.IngestorDuplicates.With(prometheus.Labels{"type": "metric", "kind": "writes_to_db"}).Inc()
```
Comment (alejandrodnm) [inline pkg/pgmodel/ingestor/duplicates.go:31]: I see that this series with `kind=writes_to_db` is not on the refactored code. Was it left out on purpose?

Comment (arajkumar) [reply pkg/pgmodel/ingestor/duplicates.go:31]: Yes, it is just a duplicate of `kind="metrics"`.

**Review by alejandrodnm — APPROVED** [review]

File: docs/runbooks/PromscaleIngestHighDataDuplication.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+There will be a high volume of samples gets into Promscale when ingesting
+from HA Prometheus. It can be examined by running the following PromQL
+query in Grafana/Prometheus.
+
+```promql
+rate(promscale_ingest_duplicates_total{kind="sample"}[5m])
+```
+
+If more data points are seen as a result of the above query, follow
```
Comment (Harkishen-Singh) [inline docs/runbooks/PromscaleIngestHighDataDuplication.md:22]: I dont think this is right. It sounds like duplicates happen when using HA improperly. While that is a case, majority of duplicates are due to timeout for Prometheus remote-write, and not HA.

I will suggest to have something like this here
```
If you see high rate of duplicate samples, check Prometheus logs for timeout or batch retry errors. If found, refer to `Tune Prometheus remote-write config`.
```

Then, we need to decide on a call, what remote-write config should we suggest for mitigation. Lets keep this as standup agenda for next meeting.

After this, we should mention the edge case, which is the HA. This might be

```
Duplicate samples can also occur on a wrongly configured HA setup. If you are using Prometheus as HA, please refer to `Prometheus high availability` for mitigation.
```

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]

File: docs/mixin/dashboards/promscale.json [hunk 4]
```diff
@@ -1774,7 +1774,7 @@
             "uid": "${DS_PROMETHEUS}"
           },
           "exemplar": true,
-          "expr": "1 - promscale_sql_database_health_check_errors_total / promscale_sql_database_health_check_total",
+          "expr": "1 - rate(promscale_sql_database_health_check_errors_total[$__rate_interval] / promscale_sql_database_health_check_total[$health_check_errors_total]",
```
Comment (paulfantom) [inline docs/mixin/dashboards/promscale.json:1777]: Where is `$health_check_errors_total` defined?

Comment (arajkumar) [reply docs/mixin/dashboards/promscale.json:1777]: This looks like a copy/paster error :)

File: docs/mixin/dashboards/promscale.json [hunk 5]
```diff
@@ -1774,7 +1774,7 @@
             "uid": "${DS_PROMETHEUS}"
           },
           "exemplar": true,
-          "expr": "1 - promscale_sql_database_health_check_errors_total / promscale_sql_database_health_check_total",
+          "expr": "1 - rate(promscale_sql_database_health_check_errors_total[$__rate_interval] / promscale_sql_database_health_check_total[$__rate_interval]",
```
Comment (alejandrodnm) [inline docs/mixin/dashboards/promscale.json:1777]:
```suggestion
          "expr": "1 - rate(promscale_sql_database_health_check_errors_total[$__rate_interval] / promscale_sql_database_health_check_total[$__rate_interval])",
```

**Comment by paulfantom** [comment]
@arajkumar what's the status of this?

**Comment by arajkumar** [comment]
> @arajkumar what's the status of this?

Thanks for the nudge @paulfantom. I will resume this and address @Harkishen-Singh's concerns.

File: docs/mixin/dashboards/promscale.json [hunk 6]
```diff
@@ -2513,7 +2513,7 @@
               },
               "editorMode": "code",
               "exemplar": false,
-              "expr": "max(promscale_sql_database_worker_maintenance_job_long_running_total{namespace=~\"$namespace\"})",
+              "expr": "max(promscale_sql_database_worker_long_running_maintenance_jobs{namespace=~\"$namespace\"})",
```
Comment (paulfantom) [inline docs/mixin/dashboards/promscale.json:2516]: Why change from `promscale_sql_database_worker_maintenance_job_long_running_total` to
`promscale_sql_database_worker_long_running_maintenance_jobs` and not to `promscale_sql_database_worker_maintenance_job_long_running`?

IMHO it is better to keep a consistent prefix of `promscale_sql_database_worker_maintenance_` to allow easier query (think `{__name__=~"promscale_sql_database_worker_maintenance_.*"}` as well as easier code navigation as `promscale_sql_database_worker_maintenance_job` could be defined in the same place regardless of how many metric registries are there.
That said I am not tied to having it either way.

**Comment by arajkumar** [comment]
@paulfantom @Harkishen-Singh PTAL.

File: pkg/util/throughput/throughput.go [hunk 7]
```diff
@@ -85,6 +91,12 @@ func (tc *throughputCalc) run() {
 			throughput = append(throughput, []interface{}{"metric-metadata/sec", int(metadataRate)}...)
 		}
 
+		duplicateSamplesRate := tc.duplicateSamples.Rate()
```
[the remark this answers is not in the document]
Comment (Harkishen-Singh) [reply pkg/util/throughput/throughput.go:94]: I am not confident of printing this as regular log. Duplicates are part of "warn" which is why we had the previous way of logging duplicates via a Rate controlled log.Warn

Having this part of throughput seems like duplicates is a feature :)

Maybe this is not a big deal. If anyone else has other thoughts, then please chime in. Otherwise I will just approve.

**Review by Harkishen-Singh — APPROVED** [review]

File: pkg/util/throughput/throughput.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (arajkumar) [reply pkg/util/throughput/throughput.go:94]: Yeah, it is not a feature, but when promscale users shares a log snippet we can easily find whether they have duplicate ingestion issue without going through the full log. Also dupicate message won't be append during a normal case!

**Review by paulfantom — APPROVED** [review]
